Public/Src/Cache/ContentStore/Library/Service/Grpc/GrpcCopyClientCache.cs

```diff
@@ -0,0 +1,146 @@
+            }
+        }
+
+        private async Task BackgroundCleanupAsync()
+        {
+            var ct = _backgroundCleaningTaskTokenSource.Token;
+
+            while (!ct.IsCancellationRequested)
+            {
+                var earliestLastUseTime = DateTime.UtcNow - TimeSpan.FromMinutes(_maximumAgeInMinutes);
+
+                var shutdownTasks = new List<Task<BoolResult>>();
+                foreach (var kvp in _clientDict)
+                {
+                    var client = kvp.Value;
+                    if (client._uses == 0 && client._lastUseTime < earliestLastUseTime)
+                    {
+                        lock (_backgroundCleanupLock)
+                        {
+                            if (client._uses == 0
+                                && client._lastUseTime < earliestLastUseTime
+                                && _clientDict.TryRemove(client.Key, out GrpcCopyClient removedClient))
+                            {
+                                // Cannot await within a lock
+                                shutdownTasks.Add(client.ShutdownAsync(_context));
+                            }
+                        }
+                    }
+                }
+
+                var allTasks = Task.WhenAll(shutdownTasks.ToArray());
+                try
+                {
+                    await allTasks;
+                }
+#pragma warning disable ERP022 // Unobserved exception in generic exception handler
+                catch (Exception)
+                {
+                    // If Task.WhenAll throws in an await, it unwraps the AggregateException and only
+                    // throws the first inner exception. We want to see all failed shutdowns.
+                    _context.Error($"Shutdown of unused GRPC clients failed after removal from client cache. {allTasks.Exception}");
+                }
+#pragma warning restore ERP022 // Unobserved exception in generic exception handler
+
+                await Task.Delay(TimeSpan.FromMinutes(_waitBetweenCleanupInMinutes), ct);
+            }
+        }
+
+        /// <summary>
+        /// Use an existing GRPC client if possible, else create a new one.
+        /// </summary>
+        /// <param name="host">Name of the host for the server (e.g. 'localhost').</param>
+        /// <param name="grpcPort">GRPC port on the server.</param>
+        /// <param name="useCompression">Whether or not GZip is enabled for copies.</param>
+        public GrpcCopyClient Create(string host, int grpcPort, bool useCompression = false)
+        {
+            var key = new GrpcCopyClientKey(host, grpcPort, useCompression);
+            if (_clientDict.TryGetValue(key, out GrpcCopyClient existingClient))
+            {
+                Interlocked.Increment(ref existingClient._uses);
+                existingClient._lastUseTime = DateTime.UtcNow;
+                return existingClient;
+            }
+            else if (_clientDict.Count > _maxClientCount)

```

Comment:
I think we should try to evict not used connections first and only after that to fail.